src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/CodeActions/Razor/SimplifyTagToSelfClosingCodeActionProvider.cs

```diff
@@ -0,0 +1,138 @@
+        }
+
+        // Make sure we're in the right kind and part of file
+        if (!FileKinds.IsComponent(context.CodeDocument.FileKind))
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        if (context.LanguageKind != RazorLanguageKind.Html)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        // Caret must be inside a markup element
+        if (context.ContainsDiagnostic(ComponentDiagnosticFactory.UnexpectedMarkupElement.Id) ||
+            context.ContainsDiagnostic(ComponentDiagnosticFactory.UnexpectedClosingTag.Id))
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var syntaxTree = context.CodeDocument.GetSyntaxTree();
+        if (syntaxTree?.Root is null)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var owner = syntaxTree.Root.FindInnermostNode(context.StartAbsoluteIndex, includeWhitespace: false)?.FirstAncestorOrSelf<MarkupTagHelperElementSyntax>();
+        if (owner is not MarkupTagHelperElementSyntax markupElementSyntax)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        // Check whether the code action is applicable to the element
+        if (!IsApplicableTo(markupElementSyntax))
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        // Provide code action to simplify
+        var actionParams = new SimplifyTagToSelfClosingCodeActionParams
+        {
+            StartTagCloseAngleIndex = markupElementSyntax.StartTag.CloseAngle.SpanStart,
+            EndTagCloseAngleIndex = markupElementSyntax.EndTag.CloseAngle.EndPosition,
+        };
+
+        var resolutionParams = new RazorCodeActionResolutionParams()
+        {
+            TextDocument = context.Request.TextDocument,
+            Action = LanguageServerConstants.CodeActions.SimplifyTagToSelfClosingAction,
+            Language = RazorLanguageKind.Razor,
+            DelegatedDocumentUri = context.DelegatedDocumentUri,
+            Data = actionParams,
+        };
+
+        var codeAction = RazorCodeActionFactory.CreateSimplifyTagToSelfClosing(resolutionParams);
+        return Task.FromResult<ImmutableArray<RazorVSInternalCodeAction>>([codeAction]);
+    }
+
+    internal static bool IsApplicableTo(MarkupTagHelperElementSyntax markupElementSyntax)
+    {
+        // Check whether the element is self-closing
+        if (markupElementSyntax is not (
+        { EndTag.CloseAngle.IsMissing: false } and
+        { StartTag.ForwardSlash: null } and
+        { StartTag.CloseAngle.IsMissing: false } and
+        { TagHelperInfo.BindingResult.Descriptors: { IsEmpty: false } descriptors }
+        ))
+        {
+            return false;
+        }
```

Comment:
I'd prefer a single pattern match over five or six. Isn't this version equivalent?

```suggestion
        if (markupElementSyntax is not
            {
                EndTag.CloseAngle.IsMissing: false,
                StartTag: { ForwardSlash: null, CloseAngle.IsMissing: false }
                TagHelperInfo.BindingResult.Descriptors: [.. var descriptors]
            })
        {
            return false;
        }
```